Add support for async callbacks in ivalue::Future

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48790 Add support for async callbacks in ivalue::Future**
* #48840 Remove DataPtr extractor from CUDAFuture
* #49014 Drop FutureNCCL in favor of vanilla CUDAFuture

This PR is a _proposal_ of a mechanism to make it easier to correctly attach an async callback to a future, without falling into blocking behaviors which could lead to deadlocks. An "async callback" is a callback that performs further network requests (e.g., on ProcessGroups, or on RPC agents) which return a future, and the _eventual result_ of such a future should be the return value of the callback. The "naive" way of making this work is having the callback wait on the inner future. However, when the callback is run by an event loop, waiting means blocking the event loop, which could prevent the inner future from ever being served, which thus leads to deadlocks.

We already fell into such issues, which I've reported in #48478. For now such instances are harmless (because of the peculiarity that NCCL doesn't have an event loop and instead always invokes callbacks inline), but we should proactively fix them.

While these issues can be fixed on the user's side (see the snippet in #48478), it requires the user to be aware of them. We should make it harder for people to fall into the wrong behavior. Thus my proposal is to allow callbacks to return the inner future directly, as it is (unawaited), rather than waiting for it. PyTorch will then take care of correctly chaining callbacks, without ever explicitly waiting.

<digression>
For those who are familiar with functional programming, this topic heavily relates to [monads](https://en.wikipedia.org/wiki/Monad_(functional_programming))! If you look at the Wikipedia page I linked, what they call `M a` we should read as `Future[TypeA]` (i.e., a Future holding a value of type `a`). The previous behavior of `then()` was the same as the `fmap` function, i.e., it takes a `M a` (`this`, i.e., the future on which one calls `then()`) and a `a -> b` (i.e., a callback that returns a value) and produces `M b` (the child future). The behavior I am adding now is the one of the `bind` function, which also takes a `M a` but its callback is of type `a -> M b` (i.e., a callback that returns a future), still with the same return type `M b`.

I don't know if this link with monads is of any relevance, but I found it somewhat insightful. It "confirms" that this new behavior of `then()` is not something crazy out of the blue but it fits in a theoretical framework. In fact, this suggests that perhaps we should not combine these two separate functions in a single "overloaded" method, but maybe leave them as separate methods (e.g., `then` and `thenAsync`? or `thenFut`?).
</digression>

Differential Revision: [D25303487](https://our.internmc.facebook.com/intern/diff/D25303487/)

lw committed:
Add support for async callbacks in ivalue::Future

This PR is a _proposal_ of a mechanism to make it easier to correctly attach an async callback to a future, without falling into blocking behaviors which could lead to deadlocks. An "async callback" is a callback that performs further network requests (e.g., on ProcessGroups, or on RPC agents) which return a future, and the _eventual result_ of such a future should be the return value of the callback. The "naive" way of making this work is having the callback wait on the inner future. However, when the callback is run by an event loop, waiting means blocking the event loop, which could prevent the inner future from ever being served, which thus leads to deadlocks.

We already fell into such issues, which I've reported in #48478. For now such instances are harmless (because of the peculiarity that NCCL doesn't have an event loop and instead always invokes callbacks inline), but we should proactively fix them.

While these issues can be fixed on the user's side (see the snippet in #48478), it requires the user to be aware of them. We should make it harder for people to fall into the wrong behavior. Thus my proposal is to allow callbacks to return the inner future directly, as it is (unawaited), rather than waiting for it. PyTorch will then take care of correctly chaining callbacks, without ever explicitly waiting.

<digression>
For those who are familiar with functional programming, this topic heavily relates to [monads](https://en.wikipedia.org/wiki/Monad_(functional_programming))! If you look at the Wikipedia page I linked, what they call `M a` we should read as `Future[TypeA]` (i.e., a Future holding a value of type `a`). The previous behavior of `then()` was the same as the `fmap` function, i.e., it takes a `M a` (`this`, i.e., the future on which one calls `then()`) and a `a -> b` (i.e., a callback that returns a value) and produces `M b` (the child future). The behavior I am adding now is the one of the `bind` function, which also takes a `M a` but its callback is of type `a -> M b` (i.e., a callback that returns a future), still with the same return type `M b`.

I don't know if this link with monads is of any relevance, but I found it somewhat insightful. It "confirms" that this new behavior of `then()` is not something crazy out of the blue but it fits in a theoretical framework. In fact, this suggests that perhaps we should not combine these two separate functions in a single "overloaded" method, but maybe leave them as separate methods (e.g., `then` and `thenAsync`? or `thenFut`?).
</digression>

Differential Revision: [D25303487](https://our.internmc.facebook.com/intern/diff/D25303487/)

[ghstack-poisoned]

lw committed:
Update on "Add support for async callbacks in ivalue::Future"


This PR is a _proposal_ of a mechanism to make it easier to correctly attach an async callback to a future, without falling into blocking behaviors which could lead to deadlocks. An "async callback" is a callback that performs further network requests (e.g., on ProcessGroups, or on RPC agents) which return a future, and the _eventual result_ of such a future should be the return value of the callback. The "naive" way of making this work is having the callback wait on the inner future. However, when the callback is run by an event loop, waiting means blocking the event loop, which could prevent the inner future from ever being served, which thus leads to deadlocks.

We already fell into such issues, which I've reported in #48478. For now such instances are harmless (because of the peculiarity that NCCL doesn't have an event loop and instead always invokes callbacks inline), but we should proactively fix them.

While these issues can be fixed on the user's side (see the snippet in #48478), it requires the user to be aware of them. We should make it harder for people to fall into the wrong behavior. Thus my proposal is to allow callbacks to return the inner future directly, as it is (unawaited), rather than waiting for it. PyTorch will then take care of correctly chaining callbacks, without ever explicitly waiting.

<digression>
For those who are familiar with functional programming, this topic heavily relates to [monads](https://en.wikipedia.org/wiki/Monad_(functional_programming))! If you look at the Wikipedia page I linked, what they call `M a` we should read as `Future[TypeA]` (i.e., a Future holding a value of type `a`). The previous behavior of `then()` was the same as the `fmap` function, i.e., it takes a `M a` (`this`, i.e., the future on which one calls `then()`) and a `a -> b` (i.e., a callback that returns a value) and produces `M b` (the child future). The behavior I am adding now is the one of the `bind` function, which also takes a `M a` but its callback is of type `a -> M b` (i.e., a callback that returns a future), still with the same return type `M b`.

I don't know if this link with monads is of any relevance, but I found it somewhat insightful. It "confirms" that this new behavior of `then()` is not something crazy out of the blue but it fits in a theoretical framework. In fact, this suggests that perhaps we should not combine these two separate functions in a single "overloaded" method, but maybe leave them as separate methods (e.g., `then` and `thenAsync`? or `thenFut`?).
</digression>

Differential Revision: [D25303487](https://our.internmc.facebook.com/intern/diff/D25303487/)

[ghstack-poisoned]

lw committed:
Update on "Add support for async callbacks in ivalue::Future"


This PR is a _proposal_ of a mechanism to make it easier to correctly attach an async callback to a future, without falling into blocking behaviors which could lead to deadlocks. An "async callback" is a callback that performs further network requests (e.g., on ProcessGroups, or on RPC agents) which return a future, and the _eventual result_ of such a future should be the return value of the callback. The "naive" way of making this work is having the callback wait on the inner future. However, when the callback is run by an event loop, waiting means blocking the event loop, which could prevent the inner future from ever being served, which thus leads to deadlocks.

We already fell into such issues, which I've reported in #48478. For now such instances are harmless (because of the peculiarity that NCCL doesn't have an event loop and instead always invokes callbacks inline), but we should proactively fix them.

While these issues can be fixed on the user's side (see the snippet in #48478), it requires the user to be aware of them. We should make it harder for people to fall into the wrong behavior. Thus my proposal is to allow callbacks to return the inner future directly, as it is (unawaited), rather than waiting for it. PyTorch will then take care of correctly chaining callbacks, without ever explicitly waiting.

<digression>
For those who are familiar with functional programming, this topic heavily relates to [monads](https://en.wikipedia.org/wiki/Monad_(functional_programming))! If you look at the Wikipedia page I linked, what they call `M a` we should read as `Future[TypeA]` (i.e., a Future holding a value of type `a`). The previous behavior of `then()` was the same as the `fmap` function, i.e., it takes a `M a` (`this`, i.e., the future on which one calls `then()`) and a `a -> b` (i.e., a callback that returns a value) and produces `M b` (the child future). The behavior I am adding now is the one of the `bind` function, which also takes a `M a` but its callback is of type `a -> M b` (i.e., a callback that returns a future), still with the same return type `M b`.

I don't know if this link with monads is of any relevance, but I found it somewhat insightful. It "confirms" that this new behavior of `then()` is not something crazy out of the blue but it fits in a theoretical framework. In fact, this suggests that perhaps we should not combine these two separate functions in a single "overloaded" method, but maybe leave them as separate methods (e.g., `then` and `thenAsync`? or `thenFut`?).
</digression>

Differential Revision: [D25303487](https://our.internmc.facebook.com/intern/diff/D25303487/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -414,7 +414,15 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
           try {
-            fut->markCompleted(cb());
+            IValue res = cb();
+            if (res.isFuture()) {
```
>  In fact, this suggests that perhaps we should not combine these two separate functions in a single "overloaded" method, but maybe leave them as separate methods (e.g., then and thenAsync? or thenFut?).

agree on the point that we should clearly distinguish these behaviors instead of packing them into the same API. We actually already have a similar thing called `@async_execution` in RPC. I wonder if we should reuse the same decorator?

https://github.com/pytorch/pytorch/blob/dee82ef3ea1ed2cae4e891f071adff03cb4a6e0d/torch/distributed/rpc/functions.py#L4

lw committed:
Update on "Add support for async callbacks in ivalue::Future"


This PR is a _proposal_ of a mechanism to make it easier to correctly attach an async callback to a future, without falling into blocking behaviors which could lead to deadlocks. An "async callback" is a callback that performs further network requests (e.g., on ProcessGroups, or on RPC agents) which return a future, and the _eventual result_ of such a future should be the return value of the callback. The "naive" way of making this work is having the callback wait on the inner future. However, when the callback is run by an event loop, waiting means blocking the event loop, which could prevent the inner future from ever being served, which thus leads to deadlocks.

We already fell into such issues, which I've reported in #48478. For now such instances are harmless (because of the peculiarity that NCCL doesn't have an event loop and instead always invokes callbacks inline), but we should proactively fix them.

While these issues can be fixed on the user's side (see the snippet in #48478), it requires the user to be aware of them. We should make it harder for people to fall into the wrong behavior. Thus my proposal is to allow callbacks to return the inner future directly, as it is (unawaited), rather than waiting for it. PyTorch will then take care of correctly chaining callbacks, without ever explicitly waiting.

<digression>
For those who are familiar with functional programming, this topic heavily relates to [monads](https://en.wikipedia.org/wiki/Monad_(functional_programming))! If you look at the Wikipedia page I linked, what they call `M a` we should read as `Future[TypeA]` (i.e., a Future holding a value of type `a`). The previous behavior of `then()` was the same as the `fmap` function, i.e., it takes a `M a` (`this`, i.e., the future on which one calls `then()`) and a `a -> b` (i.e., a callback that returns a value) and produces `M b` (the child future). The behavior I am adding now is the one of the `bind` function, which also takes a `M a` but its callback is of type `a -> M b` (i.e., a callback that returns a future), still with the same return type `M b`.

I don't know if this link with monads is of any relevance, but I found it somewhat insightful. It "confirms" that this new behavior of `then()` is not something crazy out of the blue but it fits in a theoretical framework. In fact, this suggests that perhaps we should not combine these two separate functions in a single "overloaded" method, but maybe leave them as separate methods (e.g., `then` and `thenAsync`? or `thenFut`?).
</digression>

Differential Revision: [D25303487](https://our.internmc.facebook.com/intern/diff/D25303487/)

[ghstack-poisoned]

lw committed:
Update on "Add support for async callbacks in ivalue::Future"


This PR is a _proposal_ of a mechanism to make it easier to correctly attach an async callback to a future, without falling into blocking behaviors which could lead to deadlocks. An "async callback" is a callback that performs further network requests (e.g., on ProcessGroups, or on RPC agents) which return a future, and the _eventual result_ of such a future should be the return value of the callback. The "naive" way of making this work is having the callback wait on the inner future. However, when the callback is run by an event loop, waiting means blocking the event loop, which could prevent the inner future from ever being served, which thus leads to deadlocks.

We already fell into such issues, which I've reported in #48478. For now such instances are harmless (because of the peculiarity that NCCL doesn't have an event loop and instead always invokes callbacks inline), but we should proactively fix them.

While these issues can be fixed on the user's side (see the snippet in #48478), it requires the user to be aware of them. We should make it harder for people to fall into the wrong behavior. Thus my proposal is to allow callbacks to return the inner future directly, as it is (unawaited), rather than waiting for it. PyTorch will then take care of correctly chaining callbacks, without ever explicitly waiting.

<digression>
For those who are familiar with functional programming, this topic heavily relates to [monads](https://en.wikipedia.org/wiki/Monad_(functional_programming))! If you look at the Wikipedia page I linked, what they call `M a` we should read as `Future[TypeA]` (i.e., a Future holding a value of type `a`). The previous behavior of `then()` was the same as the `fmap` function, i.e., it takes a `M a` (`this`, i.e., the future on which one calls `then()`) and a `a -> b` (i.e., a callback that returns a value) and produces `M b` (the child future). The behavior I am adding now is the one of the `bind` function, which also takes a `M a` but its callback is of type `a -> M b` (i.e., a callback that returns a future), still with the same return type `M b`.

I don't know if this link with monads is of any relevance, but I found it somewhat insightful. It "confirms" that this new behavior of `then()` is not something crazy out of the blue but it fits in a theoretical framework. In fact, this suggests that perhaps we should not combine these two separate functions in a single "overloaded" method, but maybe leave them as separate methods (e.g., `then` and `thenAsync`? or `thenFut`?).
</digression>

Differential Revision: [D25303487](https://our.internmc.facebook.com/intern/diff/D25303487/)

[ghstack-poisoned]
